recipes-extended/xen/files/xsm-for-vtd.patch

```diff
@@ -0,0 +1,259 @@
+@@ -1243,6 +1279,7 @@ static int flask_assign_device(struct domain *d, uint32_t machine_bdf)
+     u32 dsid, rsid;
+     int rc = -EPERM;
+     struct avc_audit_data ad;
++    u32 dperm = flask_iommu_resource_use_perm();
+ 
+     rc = current_has_perm(d, SECCLASS_RESOURCE, RESOURCE__ADD);
+     if ( rc )
+@@ -1259,7 +1296,7 @@ static int flask_assign_device(struct domain *d, uint32_t machine_bdf)
+         return rc;
+ 
+     dsid = domain_sid(d);
+-    return avc_has_perm(dsid, rsid, SECCLASS_RESOURCE, RESOURCE__USE, &ad);
++    return avc_has_perm(dsid, rsid, SECCLASS_RESOURCE, dperm, &ad);
+ }
+ 
+ static int flask_deassign_device(struct domain *d, uint32_t machine_bdf)
+@@ -1298,6 +1335,7 @@ static int flask_assign_dtdevice(struct domain *d, const char *dtpath)
+     u32 dsid, rsid;
+     int rc = -EPERM;
+     struct avc_audit_data ad;
++    u32 dperm = flask_iommu_resource_use_perm();
+ 
+     rc = current_has_perm(d, SECCLASS_RESOURCE, RESOURCE__ADD);
+     if ( rc )
+@@ -1314,7 +1352,7 @@ static int flask_assign_dtdevice(struct domain *d, const char *dtpath)
+         return rc;
+ 
+     dsid = domain_sid(d);
+-    return avc_has_perm(dsid, rsid, SECCLASS_RESOURCE, RESOURCE__USE, &ad);
++    return avc_has_perm(dsid, rsid, SECCLASS_RESOURCE, dperm, &ad);
+ }
+ 
+ static int flask_deassign_dtdevice(struct domain *d, const char *dtpath)
+@@ -1373,6 +1411,7 @@ struct ioport_has_perm_data {
+     u32 ssid;
+     u32 dsid;
+     u32 perm;
++    u32 use_perm;
+ };
+ 
+ static int _ioport_has_perm(void *v, u32 sid, unsigned long start, unsigned long end)
+@@ -1390,7 +1429,7 @@ static int _ioport_has_perm(void *v, u32 sid, unsigned long start, unsigned long
+     if ( rc )
+         return rc;
+ 
+-    return avc_has_perm(data->dsid, sid, SECCLASS_RESOURCE, RESOURCE__USE, &ad);
++    return avc_has_perm(data->dsid, sid, SECCLASS_RESOURCE, data->use_perm, &ad);
+ }
+ 
+ static int flask_ioport_permission(struct domain *d, uint32_t start, uint32_t end, uint8_t access)
+@@ -1411,6 +1450,7 @@ static int flask_ioport_permission(struct domain *d, uint32_t start, uint32_t en
+ 
+     data.ssid = domain_sid(current->domain);
+     data.dsid = domain_sid(d);
++    data.use_perm = flask_iommu_resource_use_perm();
+ 
+     return security_iterate_ioport_sids(start, end, _ioport_has_perm, &data);
+ }
+diff --git a/xen/xsm/flask/policy/access_vectors b/xen/xsm/flask/policy/access_vectors
+index b06cf16..92f4a0f 100644
+--- a/xen/xsm/flask/policy/access_vectors
++++ b/xen/xsm/flask/policy/access_vectors
+@@ -430,6 +430,12 @@ class resource
+ #  target = resource's security label
+ # also checked when using some core Xen devices (target xen_t)
+     use
++# IOMMU required before a resource will be added to a domain:
++# checked when adding a resource to a domain:
++#  source = domain which will have access to the resource
++#  target = resource's security label
++# also checked when using some core Xen devices (target xen_t)
```

Comment:
This line (copied from use) should be removed. It is only true for flask_resource_use_core, which is a part of the page offline sysctl and (correctly) only checks the "use" bit.